File: src/com/igneous/hw/main/TFTPPacket.java
package com.igneous.hw.main;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.math.BigInteger;
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

public class TFTPPacket {
    public static final int HEADER_SIZE = 4;
    public static final int MAX_PACKET_SIZE = 512 + HEADER_SIZE;
    public static final int RESET_BLOCK_NUM = 65535;

    public static final byte READ_REQUEST = 1;
    public static final byte WRITE_REQUEST = 2;
    public static final byte DATA = 3;
    public static final byte ACKNOWLEDGEMENT = 4;
    public static final byte ERROR = 5;

    public static final byte ERROR_NOT_DEFINED = 0;
    public static final byte ERROR_FILE_NOT_FOUND = 1;
    public static final byte ERROR_DISK_FULL = 3;
    public static final byte ERROR_ILLEGAL_TFTP_OPERATION = 4;
    public static final byte ERROR_UNKNOWN_TRANSFER_ID = 5;
    public static final byte ERROR_FILE_ALREADY_EXISTS = 6;
    public static final Map<Byte, String> ERROR_MESSAGES = new HashMap<>(3);
    static {
        ERROR_MESSAGES.put(ERROR_NOT_DEFINED, "Error not defined");
        ERROR_MESSAGES.put(ERROR_FILE_NOT_FOUND, "File not found");
        ERROR_MESSAGES.put(ERROR_ILLEGAL_TFTP_OPERATION, "Illegal TFTP operation");
        ERROR_MESSAGES.put(ERROR_UNKNOWN_TRANSFER_ID, "Unknown transfer id");
        ERROR_MESSAGES.put(ERROR_FILE_ALREADY_EXISTS, "File already exists");
        ERROR_MESSAGES.put(ERROR_DISK_FULL, "Disk full or allocation exceeded");
    }
    public static final byte NULL_TERMINATOR = 0;
    public static final String MODE = "octet";

    private static final Logger LOGGER = Logger.getLogger(TFTPPacket.class.getSimpleName());
    private byte[] opCode = new byte[2];
    private byte[] blockNumber = new byte[2];
    private byte[] errorCode = new byte[2];
    private String filename;
    private byte[] data;
    private int length;
    private int type;
    private int port;
    private InetAddress address;
    private String errorMessage;

    public TFTPPacket(int type, InetAddress address, int port) {
        this.type = type;
        this.address = address;
        this.port = port;
    }

    public int getLength() {
        return length;
    }

    public void setLength(int length) {
        this.length = length;
    }

    public void setErrorCode(byte eCode) {
        this.errorCode[0] = 0;
        switch (eCode) {
            case ERROR_NOT_DEFINED:
            case ERROR_FILE_NOT_FOUND:
            case ERROR_DISK_FULL:
            case ERROR_FILE_ALREADY_EXISTS:
            case ERROR_ILLEGAL_TFTP_OPERATION:
            case ERROR_UNKNOWN_TRANSFER_ID:
                this.errorCode[1] = eCode;
                break;
            default:
                this.errorCode[1] = -1;
        }
    }

    public void setErrorMessage(String errorMessage) {
        this.errorMessage = errorMessage;
    }

    public int getType() {
        return type;
    }

    public InetAddress getAddress() {
        return address;
    }

    public int getPort() {
        return port;
    }

    public String getFilename() {
        return filename;
    }

    public void setFilename(String filename) {
        this.filename = filename;
    }

    public byte[] getBlockNumber() {
        return blockNumber;
    }

    public void setBlockNumber(byte[] blockNumber) {
        this.blockNumber = blockNumber;
    }

    public byte[] getData() {
        return data;
    }

    public void setData(byte[] data) {
        this.data = data;
    }

    public DatagramPacket buildDatagram() {
        try {
            byte[] buffer = new byte[length + HEADER_SIZE];
            ByteArrayOutputStream bos = new ByteArrayOutputStream();
            DatagramPacket sendPacket = new DatagramPacket(buffer, buffer.length, address, port);
            opCode[0] = 0;
            opCode[1] = BigInteger.valueOf(type).toByteArray()[0];
            bos.write(opCode, 0, 2);
            switch (opCode[1]) {
                case READ_REQUEST:
                case WRITE_REQUEST:
                    bos.write(filename.getBytes());
                    bos.write(NULL_TERMINATOR);
                    bos.write(MODE.getBytes());
                    bos.write(NULL_TERMINATOR);
                    sendPacket.setData(bos.toByteArray());
                    return sendPacket;
                case DATA:
                    bos.write(blockNumber);
                    bos.write(data);
                    sendPacket.setData(bos.toByteArray());
                    return sendPacket;
                case ACKNOWLEDGEMENT:
                    bos.write(blockNumber);
                    sendPacket.setData(bos.toByteArray());
                    return sendPacket;
                case ERROR:
                    bos.write(errorCode);
                    bos.write(errorMessage.getBytes());
                    bos.write(NULL_TERMINATOR);
                    sendPacket.setData(bos.toByteArray());
                    return sendPacket;
                default:
                    return null;
            }
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, e.getMessage(), e);
        }
        return null;
    }

    public static TFTPPacket parseTFTPPacket(DatagramPacket datagramPacket) {
        try {
            ByteArrayInputStream bis = new ByteArrayInputStream(datagramPacket.getData(), 0, datagramPacket.getData().length);
            byte[] opCode = new byte[2];
            bis.read(opCode);
            TFTPPacket packet = new TFTPPacket(opCode[1], datagramPacket.getAddress(), datagramPacket.getPort());
            switch (opCode[1]) {
                case READ_REQUEST:
                case WRITE_REQUEST:
                    byte[] arr = new byte[bis.available()];
                    byte next = (byte) bis.read();
                    int count = 0;
                    while (next != NULL_TERMINATOR) {
                        arr[count++] = next;
                        next = (byte) bis.read();
                    }
                    packet.setFilename(new String(arr));
                    return packet;
                case DATA:
                    bis.read(packet.blockNumber);
                    packet.setLength(datagramPacket.getLength());
                    packet.data = new byte[bis.available()];
                    bis.read(packet.data);
                    return packet;
                case ACKNOWLEDGEMENT:
                    bis.read(packet.blockNumber);
                    return packet;
                case ERROR:
                    bis.read(packet.errorCode);
                    packet.setLength(datagramPacket.getLength());
                    packet.data = new byte[bis.available() - 1];
                    bis.read(packet.data);
                    String errorMessage = new String(packet.data);
                    packet.setErrorMessage(errorMessage);
                    return packet;
                default:
                    return null;
            }
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, e.getMessage(), e);
        }
        return null;
    }
}

File: go/src/packet/TFTPPacket.go
package packet

import (
	"bytes"
	"encoding/binary"
)

const HEADER_SIZE = 4
const MAX_PACKET_SIZE uint16 = 512 + HEADER_SIZE
const RESET_BLOCK_NUM uint16 = 65535

const READ_REQUEST uint16 = 1
const WRITE_REQUEST uint16 = 2
const DATA uint16 = 3
const ACKNOWLEDGEMENT uint16 = 4
const ERROR uint16 = 5

const ERROR_NOT_DEFINED uint16 = 0
const ERROR_FILE_NOT_FOUND uint16 = 1
const ERROR_DISK_FULL uint16 = 3
const ERROR_ILLEGAL_TFTP_OPERATION uint16 = 4
const ERROR_UNKNOWN_TRANSFER_ID uint16 = 5
const ERROR_FILE_ALREADY_EXISTS uint16 = 6

var ERROR_MESSAGES = map[uint16]string{
	ERROR_NOT_DEFINED :  "Error not defined",
	ERROR_FILE_NOT_FOUND :  "File not found",
	ERROR_ILLEGAL_TFTP_OPERATION :  "Illegal TFTP operation",
	ERROR_UNKNOWN_TRANSFER_ID :  "Unknown transfer id",
	ERROR_FILE_ALREADY_EXISTS : "File already exists",
	ERROR_DISK_FULL :  "Disk full or allocation exceeded"}

const NULL_TERMINATOR byte = 0
const MODE string = "octet"

var blockNumber uint16
var errorCode uint16
var filename string
var data []byte
var length int
var packetType uint16
var port uint16
var address string
var errorMessage string

type TFTPPacket struct {
	packetType uint16
	address    string
	port       uint16
}

func makeTFTPPacket(packetType1 uint16, address1 string, port1 uint16) TFTPPacket {
	packetType = packetType1
	address = address1
	port = port1
	return TFTPPacket{packetType1, address1, port1}
}

func getLength() int {
	return length
}

func setLength(length1 int) {
	length = length1
}

func setErrorCode(eCode uint16) {
	switch (eCode) {
	case ERROR_NOT_DEFINED:
	case ERROR_FILE_NOT_FOUND:
	case ERROR_DISK_FULL:
	case ERROR_FILE_ALREADY_EXISTS:
	case ERROR_ILLEGAL_TFTP_OPERATION:
	case ERROR_UNKNOWN_TRANSFER_ID:
		errorCode = eCode
		break
	default:
		errorCode = 0
	}
}

func setErrorMessage(errorMessage1 string) {
	errorMessage = errorMessage1
}

func getType() uint16 {
	return packetType
}

func setAddress(address1 string) {
	address = address1
}

func getAddress() string {
	return address
}

func setPort(port1 uint16) {
	port = port1
}

func getPort() uint16 {
	return port
}

func getFilename() string {
	return filename
}

func setFilename(filename1 string) {
	filename = filename1
}

func getBlockNumber() uint16 {
	return blockNumber
}

func setBlockNumber(blockNumber1 uint16) {
	blockNumber = blockNumber1
}

func getData() []byte {
	return data
}

func setData(data []byte) {
	data = data
}

func buildDatagram() []byte {
	buffer := new(bytes.Buffer)
	binary.Write(buffer, binary.BigEndian, packetType)
	switch (packetType) {
	case READ_REQUEST:
	case WRITE_REQUEST:
		buffer.WriteString(filename)
		buffer.WriteByte(NULL_TERMINATOR)
		buffer.WriteString(MODE)
		buffer.WriteByte(NULL_TERMINATOR)
		return buffer.Bytes()
	case DATA:
		binary.Write(buffer, binary.BigEndian, blockNumber)
		buffer.Write(data)
		return buffer.Bytes()
	case ACKNOWLEDGEMENT:
		binary.Write(buffer, binary.BigEndian, blockNumber)
		return buffer.Bytes()
	case ERROR:
		binary.Write(buffer, binary.BigEndian, errorCode)
		buffer.WriteString(errorMessage)
		buffer.WriteByte(NULL_TERMINATOR)
		return buffer.Bytes()
	default:
		return nil
	}
	return nil
}

func parseTFTPPacket(buf []byte) TFTPPacket {
	packetType := binary.BigEndian.Uint16(buf[0:2])
	packet := makeTFTPPacket(packetType, getAddress(), getPort())
	switch (packetType) {
	case READ_REQUEST:
	case WRITE_REQUEST:
		filename := bytes.Split(buf[2:], []byte{0})[0]
		setFilename(string(filename))
		return packet
	case DATA:
		blockNumber := binary.BigEndian.Uint16(buf[2:4])
		setBlockNumber(blockNumber)
		setLength(len(buf[4:]))
		data := buf[4:]
		setData(data)
		return packet
	case ACKNOWLEDGEMENT:
		blockNumber := binary.BigEndian.Uint16(buf[2:4])
		setBlockNumber(blockNumber)
		return packet
	case ERROR:
		errorCode := binary.BigEndian.Uint16(buf[2:4])
		setErrorCode(errorCode)
		setLength(len(buf[4:]))
		errorMessage := string(buf[4:])
		setErrorMessage(errorMessage)
		return packet
	default:
		return TFTPPacket{}
	}
	return TFTPPacket{}
}
